wts/blueprints/aggregate.py

```diff
@@ -0,0 +1,115 @@
+import asyncio
+import flask
+import httpx
+import time
+from cdiserrors import NotFoundError, UserError
+
+from ..auth import authenticate
+from ..models import db, RefreshToken
+from ..tokens import async_get_access_token
+
+
+blueprint = flask.Blueprint("aggregate", __name__)
+
+
+@blueprint.route("/<path:endpoint>", methods=["GET"])
+async def get_aggregate_response(endpoint):
+    """
+    Proxy GET requests to `endpoint` on each linked commons and return
+    an aggregated response.
+
+    For an authenticated request, each proxied request incudes an access token
+    fetched using the current user's refresh token.
+
+    The size of the aggregated response body can be reduced by supplying a
+    `filters` parameter. If provided, only return JSON key-value pairs whose
+    key is in `filters`. Multiple filters can be specified:
+
+    `GET /aggregate/user/user?filters=authz&filters=username`
+    """
+    # for `GET /aggregate/user/user`, flask sets endpoint to 'user/user'
+    endpoint = f"/{endpoint}"
+    if endpoint not in flask.current_app.config["AGGREGATE_ENDPOINT_ALLOWLIST"]:
+        raise NotFoundError(
+            "supplied endpoint is not configured in the Workspace Token Service aggregate endpoint allowlist"
+        )
+
+    filters = flask.request.args.getlist("filters")
+    parameters = flask.request.args.to_dict()
+    parameters.pop("filters", None)
+
+    if flask.request.headers.get("Authorization"):
+        authenticate(allow_access_token=True)
+        refresh_tokens = (
+            db.session.query(RefreshToken)
+            .filter_by(username=flask.g.user.username)
+            .filter(RefreshToken.expires > int(time.time()))
+            .order_by(RefreshToken.expires.asc())
+        )
+        #  if a user has multiple refresh tokens for the same commons, we want
+        #  the latest one to be used. see
+        #  https://stackoverflow.com/questions/39678672/is-a-python-dict-comprehension-always-last-wins-if-there-are-duplicate-keys
+        refresh_tokens = {
+            flask.current_app.config["OIDC"][rt.idp]["commons_hostname"]: rt
+            for rt in refresh_tokens
+        }
+        access_tokens = await asyncio.gather(
+            *[async_get_access_token(rt) for rt in refresh_tokens.values()]
+        )
+        request_info = [
+            (commons, endpoint, {"Authorization": f"Bearer {access_token}"})
+            if access_token
+            else (commons, None, None)
+            for commons, access_token in access_tokens
+        ]
+    else:
+        request_info = [
```

Comment:
Wait, so this is allowing making aggregate requests without authenticating at all?

Reply:
I don't think we want to allow this, it exposes a way to potentially spam all our commons at once with requests. We should ensure that only a valid BRH user is using this... and honestly we may want to consider specific arborist policy that enables someone to use this endpoint. Has the authorization around this been discussed?

Reply:
Today, when an unauthenticated user navigates to the discovery portal, that user sees an unlocked icon for open access studies. Having WTS proxy unauthenticated requests was an attempt [to preserve that end user behavior](https://github.com/uc-cdis/data-portal/pull/899#discussion_r678401838).

I'll follow up with more details regarding discussion of authorization.

Reply:
oh b/c an unauthed request to mapping provides the public access information. ahh, I see